File: 이상욱/BOJ/m3w8/BOJ1197_kruskal.java
package m3w8;

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.StringTokenizer;

public class BOJ1197_kruskal {
	static int V,E;
	static int[] p;
	static Node_[] arr;
	public static void main(String[] args) throws Exception{
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st = new StringTokenizer(br.readLine());
		
		V = Integer.parseInt(st.nextToken());
		E = Integer.parseInt(st.nextToken());
		
		arr = new Node_[E];
		p = new int[V+1];
		
		for(int i = 1; i <= V; i++) {
			p[i] = i;
		}
		
		for(int i = 0; i < E; i++) {
			st = new StringTokenizer(br.readLine());
			int a = Integer.parseInt(st.nextToken());
			int b = Integer.parseInt(st.nextToken());
			int c = Integer.parseInt(st.nextToken());
			
			arr[i] = new Node_(a,b,c);
		}
		
		System.out.println(kruskal());
	}
	private static int kruskal() {
		
		int res = 0;
		int cnt = 0;
		
		Arrays.sort(arr);
		
		for( Node_ n : arr ) {
			if(union(n.s,n.d)) {
				res += n.w;
				cnt++;
				if(cnt == V-1) {
					return res;
				}
			}
		}
		
		return res;
	}
	private static boolean union(int s, int d) {
	
		int sP = find(s);
		int dP = find(d);
		
		if(sP == dP) {
			return false;
		}
		
		p[sP] = p[dP];
		return true;
	}
	
	
	private static int find(int s) {
		
		if(s == p[s]) {
			return s;
		}
		
		return p[s] = find(p[s]);
				
	}
	
}

class Node_ implements Comparable<Node_>{
	
	int s;
	int d;
	int w;
	
	
	public Node_(int s, int d, int w) {
		this.s = s;
		this.d = d;
		this.w = w;
	}


	@Override
	public int compareTo(Node_ o) {
		
		return this.w - o.w;
	}
	
}

File: 이상욱/BOJ/m3w8/BOJ1197_kruskal.kt
package m3w8

import java.io.BufferedReader
import java.io.InputStreamReader
import java.util.*

var V = 0
var E = 0
lateinit var parent : IntArray
lateinit var arr : Array<Node>

fun main(){

    val br = BufferedReader(InputStreamReader(System.`in`))
    var st = StringTokenizer(br.readLine())

    V = st.nextToken().toInt()
    E = st.nextToken().toInt()

    arr = Array(E){Node(0,0,0)}
    parent = IntArray(V+1)

    for (i in 1..V){
        parent[i] = i
    }

    for(i in 0 until E) {
        st = StringTokenizer(br.readLine())
        val a = st.nextToken().toInt()
        val b = st.nextToken().toInt()
        val c = st.nextToken().toInt()

        arr[i] = Node(a,b,c)
    }

    println(kruskal())
}

fun kruskal() : Int{

    var res = 0
    var cnt = 0

    arr.sort()
    arr.forEach {
        if(union_(it.s,it.d)){
            res+= it.w
            cnt++
            if(cnt == V-1){
                return res
            }
        }
    }
    return res
}

fun union_(a : Int, b : Int) : Boolean{
    val aP = find(a)
    val bP = find(b)

    if(aP == bP) {
        return false
    }
    parent[aP] = parent[bP]
    return true
}

fun find(a : Int) : Int{
    if(a == parent[a]){
        return a
    }
    parent[a] = find(parent[a])
    return parent[a]
}

data class Node(val s : Int,val d : Int, val w : Int):Comparable<Node> {
    override fun compareTo(other: Node): Int {
        return w - other.w
    }
}